dkg/sync/client.go

```diff
@@ -32,168 +32,178 @@ import (
-	msg := &pb.MsgSync{
-		Timestamp:     timestamppb.Now(),
-		HashSignature: hashSig,
-		Shutdown:      false,
+// AwaitConnected blocks until the connection with the server has been established or returns a context error.
+func (c *Client) AwaitConnected(ctx context.Context) error {
+	timer := time.NewTicker(time.Millisecond)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-timer.C:
+			if c.isConnected() {
+				return nil
+			}
+		}
 	}
+}
 
-	resp, err := c.send(msg)
-	if err != nil {
-		return result{error: err}
-	}
+// Shutdown triggers the Run goroutine to shut down gracefully and returns nil after it has returned.
+// It should be called after AwaitConnected and may only be called once.
+func (c *Client) Shutdown(ctx context.Context) error {
+	close(c.shutdown)
 
-	return result{
-		rtt:       time.Since(before),
-		timestamp: resp.SyncTimestamp.String(),
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.done:
+		return nil
 	}
 }
 
-func (c *Client) send(msg *pb.MsgSync) (*pb.MsgSyncResponse, error) {
-	wb, err := proto.Marshal(msg)
-	if err != nil {
-		return nil, errors.Wrap(err, "marshal msg")
-	}
+func (c *Client) setConnected() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	if _, err = c.stream.Write(wb); err != nil {
-		return nil, errors.Wrap(err, "write msg to stream")
-	}
+	c.connected = true
+}
 
-	buf := bufio.NewReader(c.stream)
-	rb := make([]byte, MsgSize)
-	// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
-	n, err := buf.Read(rb)
-	if err != nil {
-		return nil, errors.Wrap(err, "read server response")
-	}
+func (c *Client) clearConnected() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	// The first `n` bytes that are read are the most important
-	rb = rb[:n]
+	c.connected = false
+}
 
-	resp := new(pb.MsgSyncResponse)
-	if err = proto.Unmarshal(rb, resp); err != nil {
-		return nil, errors.Wrap(err, "unmarshal server response")
-	} else if resp.Error != "" {
-		return nil, errors.New(resp.Error)
-	}
+func (c *Client) isConnected() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	return resp, nil
+	return c.connected
 }
 
-// NewClient starts a goroutine that establishes a long lived connection to a p2p server and returns a new Client instance.
-// TODO(dhruv): call onFailure on permanent failure.
-func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func()) *Client {
-	s, err := tcpNode.NewStream(ctx, server.ID, syncProtoID)
-	if err != nil {
-		log.Error(ctx, "Open new stream with server", err)
-		ch := make(chan result, 1)
-		ch <- result{error: err}
-		close(ch)
-
-		return &Client{
-			ctx:       ctx,
-			onFailure: onFailure,
-			tcpNode:   tcpNode,
-			server:    server,
-			results:   ch,
+func (c *Client) sendMsgs(ctx context.Context, stream network.Stream) (bool, error) {
+	timer := time.NewTicker(time.Second)
+	defer timer.Stop()
+
+	first := make(chan struct{}, 1)
```

Comment:
what is the significance of this channel?

Reply:
just so a message is sent initially, instead of only 1 second later